File: ios/Sendro/Core/PairedHostStore.swift
//
//  PairedHostStore.swift
//  Sendro
//
//  Persistent registry of paired Windows hosts. The bearer token itself
//  lives in the Keychain (KeychainStore); this store keeps metadata plus
//  the last known reachable endpoint.
//

import Foundation
import Combine

struct PairedHost: Codable, Identifiable, Equatable {
    let deviceId: String        // host deviceId (uuid, lowercase)
    var name: String
    var lastHost: String        // ip or hostname
    var lastPort: UInt16
    var pairedAtMs: Int64

    var id: String { deviceId }
}

final class PairedHostStore: ObservableObject {

    @Published private(set) var hosts: [PairedHost] = []

    private static let defaultsKey = "sendro.pairedHosts"

    init() {
        load()
    }

    func host(id: String) -> PairedHost? {
        hosts.first { $0.deviceId == id }
    }

    func add(_ host: PairedHost) {
        var list = hosts.filter { $0.deviceId != host.deviceId }
        list.append(host)
        hosts = list.sorted { $0.pairedAtMs < $1.pairedAtMs }
        save()
    }

    func remove(id: String) {
        hosts.removeAll { $0.deviceId == id }
        save()
    }

    /// Called when discovery resolves a fresh address for a paired host.
    func updateEndpoint(id: String, host: String, port: UInt16, name: String?) {
        guard let idx = hosts.firstIndex(where: { $0.deviceId == id }) else { return }
        var entry = hosts[idx]
        var changed = false
        if entry.lastHost != host { entry.lastHost = host; changed = true }
        if entry.lastPort != port { entry.lastPort = port; changed = true }
        if let name, !name.isEmpty, entry.name != name { entry.name = name; changed = true }
        if changed {
            hosts[idx] = entry
            save()
        }
    }

    private func load() {
        guard let data = UserDefaults.standard.data(forKey: Self.defaultsKey) else { return }
        if let decoded = try? JSONDecoder().decode([PairedHost].self, from: data) {
            hosts = decoded
        }
    }

    private func save() {
        if let data = try? JSONEncoder().encode(hosts) {
            UserDefaults.standard.set(data, forKey: Self.defaultsKey)
        }
    }
}

File: android/app/src/main/java/com/sendro/android/core/PairedHostStore.kt
package com.sendro.android.core

import android.content.Context
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.serialization.Serializable
import kotlinx.serialization.decodeFromString
import kotlinx.serialization.encodeToString

/**
 * A Windows host this phone has paired with. The bearer token itself lives in
 * [TokenStore]; this keeps the metadata plus the last endpoint that worked, so
 * a manually-typed or QR-scanned PC is reachable again after a restart with no
 * discovery at all.
 */
@Serializable
data class PairedHost(
    val deviceId: String,
    val name: String,
    val lastHost: String,
    val lastPort: Int,
    val pairedAtMs: Long,
    /**
     * `windows` | `androidtv` | `android` | `ios`. Informational only (§15.1) —
     * it decides an icon and a label, never a capability.
     *
     * Defaulted so entries written before this field existed still decode.
     */
    val platform: String = "windows",
    /**
     * True once this peer has answered `404` to an outbox poll (§15.1): it is a
     * receiver, not a sender. Cached so the poll loop stops long-polling it and
     * the UI can say "receive-only" instead of showing it as broken.
     */
    val receiveOnly: Boolean = false,
)

/**
 * Persistent registry of paired hosts. Small and read on the UI thread, so
 * plain SharedPreferences + JSON rather than DataStore — the whole list is a
 * few hundred bytes and the engine needs it synchronously.
 */
class PairedHostStore(context: Context) {

    private val prefs = context.applicationContext
        .getSharedPreferences("sendro_paired_hosts", Context.MODE_PRIVATE)

    private val _hosts = MutableStateFlow(load())
    val hosts: StateFlow<List<PairedHost>> = _hosts.asStateFlow()

    fun host(id: String?): PairedHost? =
        if (id == null) null else _hosts.value.firstOrNull { it.deviceId == id }

    fun add(host: PairedHost) {
        val next = (_hosts.value.filter { it.deviceId != host.deviceId } + host)
            .sortedBy { it.pairedAtMs }
        _hosts.value = next
        save(next)
    }

    fun remove(id: String) {
        val next = _hosts.value.filter { it.deviceId != id }
        _hosts.value = next
        save(next)
    }

    /**
     * Called when discovery (or a successful manual connect) finds a fresh
     * address for a host we already know.
     */
    /** Records the §15.1 outbox-404 discovery. */
    @Synchronized
    fun markReceiveOnly(id: String, receiveOnly: Boolean) {
        val current = _hosts.value
        val index = current.indexOfFirst { it.deviceId == id }
        if (index < 0) return
        if (current[index].receiveOnly == receiveOnly) return
        val next = current.toMutableList()
        next[index] = next[index].copy(receiveOnly = receiveOnly)
        _hosts.value = next
        save(next)
    }

    fun updateEndpoint(id: String, address: String, port: Int, name: String?) {
        val current = _hosts.value
        val index = current.indexOfFirst { it.deviceId == id }
        if (index < 0) return
        val existing = current[index]
        val updated = existing.copy(
            lastHost = address,
            lastPort = port,
            name = if (!name.isNullOrBlank()) name else existing.name,
        )
        if (updated == existing) return
        val next = current.toMutableList().also { it[index] = updated }
        _hosts.value = next
        save(next)
    }

    private fun load(): List<PairedHost> {
        val raw = prefs.getString(KEY, null) ?: return emptyList()
        return runCatching { SendroJson.decodeFromString<List<PairedHost>>(raw) }
            .getOrDefault(emptyList())
    }

    private fun save(list: List<PairedHost>) {
        prefs.edit().putString(KEY, SendroJson.encodeToString(list)).apply()
    }

    private companion object {
        const val KEY = "hosts"
    }
}
